crates/matrix-sdk/src/sliding_sync/list/mod.rs

```diff
@@ -1004,121 +822,46 @@ mod tests {
-
-            assert_eq!(*list.inner.ranges.read().unwrap(), &[0..=1]);
-
-            list.add_range(2..=3).unwrap();
-            assert_eq!(*list.inner.ranges.read().unwrap(), &[0..=1, 2..=3]);
-        }
-
-        // Add range on `Growing`.
-        {
-            let list = SlidingSyncList::builder("foo")
-                .sync_mode(SlidingSyncMode::new_growing(20, None))
-                .build(sender.clone());
-
-            assert!(list.add_range(2..=3).is_err());
-        }
+        let list = SlidingSyncList::builder("foo")
+            .sync_mode(SlidingSyncMode::new_selective().add_range(0..=1).add_range(2..=3))
+            .build(sender);
 
-        // Add range on `Paging`.
         {
-            let list = SlidingSyncList::builder("foo")
-                .sync_mode(SlidingSyncMode::new_paging(20, None))
-                .build(sender);
+            let mut generator = list.inner.request_generator.write().unwrap();
+            assert_eq!(generator.requested_ranges(), &[0..=1, 2..=3]);
 
-            assert!(list.add_range(2..=3).is_err());
+            let ranges = generator.generate_next_ranges(None).unwrap();
+            assert_eq!(ranges, &[0..=1, 2..=3]);
         }
-    }
 
-    #[test]
-    fn test_sliding_sync_list_reset_ranges() {
-        let (sender, _receiver) = channel(1);
-
-        // Reset ranges on `Selective`.
-        {
-            let list = SlidingSyncList::builder("foo")
-                .sync_mode(SlidingSyncMode::new_selective())
-                .add_range(0..=1)
-                .build(sender.clone());
+        // There shouldn't be any internal request to restart the sync loop yet.
+        assert!(matches!(receiver.try_recv(), Err(TryRecvError::Empty)));
 
-            assert_eq!(*list.inner.ranges.read().unwrap(), &[0..=1]);
+        list.set_sync_mode(SlidingSyncSelectiveModeBuilder::new().add_range(4..=5)).unwrap();
 
-            list.reset_ranges().unwrap();
-            assert!(list.inner.ranges.read().unwrap().is_empty());
-        }
-
-        // Reset range on `Growing`.
         {
-            let list = SlidingSyncList::builder("foo")
-                .sync_mode(SlidingSyncMode::new_growing(20, None))
-                .build(sender.clone());
+            let mut generator = list.inner.request_generator.write().unwrap();
+            assert_eq!(generator.requested_ranges(), &[4..=5]);
 
-            assert!(list.reset_ranges().is_err());
+            let ranges = generator.generate_next_ranges(None).unwrap();
+            assert_eq!(ranges, &[4..=5]);
         }
 
-        // Reset range on `Paging`.
-        {
-            let list = SlidingSyncList::builder("foo")
-                .sync_mode(SlidingSyncMode::new_paging(20, None))
-                .build(sender);
-
-            assert!(list.reset_ranges().is_err());
-        }
+        // Setting the sync mode requests exactly one restart of the sync loop.
+        assert!(receiver.try_recv().is_ok());
```

Comment:
Ideally, we would want to test the content of the received value so that the comment is ensured to be valid :-).

Reply:
Oh right, now we have the stop message too. Fixed